core/dbt/deps/registry.py

```diff
@@ -125,20 +127,26 @@ def resolved(self) -> RegistryPinnedPackage:
             new_msg = "Version error for package {}: {}".format(self.name, e)
             raise DependencyException(new_msg) from e
 
-        available = registry.get_available_versions(self.package)
+        should_version_check = bool(flags.VERSION_CHECK)
+        dbt_version = get_installed_version()
+        compatible_versions = registry.get_compatible_versions(
+            self.package, dbt_version, should_version_check
+        )
         prerelease_version_specified = any(bool(version.prerelease) for version in self.versions)
         installable = semver.filter_installable(
-            available, self.install_prerelease or prerelease_version_specified
+            compatible_versions, self.install_prerelease or prerelease_version_specified
         )
-        available_latest = installable[-1]
-
-        # for now, pick a version and then recurse. later on,
-        # we'll probably want to traverse multiple options
-        # so we can match packages. not going to make a difference
-        # right now.
-        target = semver.resolve_to_specific_version(range_, installable)
+        if installable:
+            # for now, pick a version and then recurse. later on,
+            # we'll probably want to traverse multiple options
+            # so we can match packages. not going to make a difference
+            # right now.
+            target = semver.resolve_to_specific_version(range_, installable)
+        else:
+            target = None
         if not target:
-            package_version_not_found(self.package, range_, installable)
+            package_version_not_found(self.package, range_, installable, should_version_check)
+        latest_compatible = installable[-1]
```

Comment:
Ah, I see now! Definitely a bit tricky. Thanks for the explanation.

Maybe add a quick comment like:
```suggestion
            # raise an exception if no installable target version is found
            package_version_not_found(self.package, range_, installable, should_version_check)
        latest_compatible = installable[-1]
```